File: packages/shared-types/src/recommendation.ts
import type { ProblemSource } from "./history";

export type RecommendationType = "weakness" | "next_step" | "review" | string;

export type RecommendationItem = {
  id: string;
  recommendation_type: RecommendationType;
  problem_source: ProblemSource;
  lesson_id: string | null;
  generated_problem_id: string | null;
  priority_score: string;
  score_breakdown: Record<string, unknown>;
  reason: string | null;
  lesson_title: string | null;
};

export type RecommendationListResponse = {
  items: RecommendationItem[];
};

export type WeaknessItem = {
  tag: string;
  total_attempts: number;
  correct_count: number;
  accuracy_rate: string;
  avg_score: string;
};

export type WeaknessProfileResponse = {
  items: WeaknessItem[];
};

File: apps/backend/app/schemas/recommendation.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel


class RecommendationItem(BaseModel):
    id: str
    recommendation_type: str
    problem_source: str
    lesson_id: str | None
    generated_problem_id: str | None
    priority_score: Decimal
    score_breakdown: dict
    reason: str | None
    lesson_title: str | None = None

    model_config = {"from_attributes": True}


class RecommendationListResponse(BaseModel):
    items: list[RecommendationItem]


class WeaknessItem(BaseModel):
    tag: str
    total_attempts: int
    correct_count: int
    accuracy_rate: Decimal
    avg_score: Decimal

    model_config = {"from_attributes": True}


class WeaknessProfileResponse(BaseModel):
    items: list[WeaknessItem]
